### PR TITLE
BCIs are now stored in the manipulation chamber's contents, instead of nullspace

### LINKED ISSUE #70349: MMI BCI Nullspace Bug
## Reproduction:
Step 1: Insert an MMI component into a BCI.
Step 2: Insert the BCI into a BCI manipulator.
Step 3: MMI goes into the nullspace room.
(optional) Step 4: MMI cries for help in the nullspace room only for no-one to hear it's cries.

### PR DESCRIPTION
**Pull request by RikuTheKiller**
## About The Pull Request

Fixes a bug where MMIs would be sent into the nullspace room when in an MMI component inside of a BCI that is inserted into a BCI manipulation chamber. 

Stores the BCI in the chamber's contents, instead of nullspace, allowing the MMI to talk, the BCI to still act as it would normally and overall just solves the issue and maybe some others alongside it.

Fixes #70349

## Why It's Good For The Game

Bug fix good.

## Changelog

:cl:
fix: Fixed a nullspacing bug related to BCIs, specifically MMI components that are contained inside of them.
refactor: Refactored some of the BCI manipulation chamber's code.
/:cl:

### CODE REVIEW

**Comment by msgerbs** [comment]
> > You currently have a negative Fix/Feature pull request delta of -5. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.
> 
> I sometimes wonder why it doesn't filter out PRs that only have fix-oriented tags.

"You have negative GBP so we're gonna close your fix that would raise your GBP!" does seem kind of mean doesn't it.

**Comment by Mothblocks** [comment]
Two reasons.

1. The GBP action runs before the labeling webhook. Fixing this is nontrivial.
2. We regularly have cases where people tag stuff as fixes when they are just balance changes. I have appreciated the bot being there.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 1]
```diff
@@ -290,21 +290,15 @@
...
 
 /obj/machinery/bci_implanter/on_deconstruction()
-	var/obj/item/organ/internal/cyberimp/bci/bci_to_implant_resolved = bci_to_implant?.resolve()
-	bci_to_implant_resolved?.forceMove(drop_location())
-	bci_to_implant = null
+	drop_stored_bci()
 
 /obj/machinery/bci_implanter/Destroy()
-	QDEL_NULL(bci_to_implant)
+	drop_stored_bci()
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:296]: Destroy() should never add things to the world. bci_implanter/Destroy() should destroy the BCI.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 2]
```diff
@@ -290,21 +290,15 @@
...
 
 /obj/machinery/bci_implanter/examine(mob/user)
-	. = ..()
-
-	if (isnull(bci_to_implant?.resolve()))
-		. += span_notice("There is no BCI inserted.")
-	else
-		. += span_notice("Right-click to remove current BCI.")
+	. += span_notice(bci_to_implant ? "Right-click to remove current BCI." : "There is no BCI inserted.")
+	return ..()
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:301]: This doesn't work and will not include the notice you added. Revert this change and replace `if (isnull(bci_to_implant?.resolve()))` with `if (isnull(bci_to_implant))`.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 3]
```diff
@@ -281,7 +281,7 @@
 	var/busy_icon_state
 	var/locked = FALSE
 
-	var/datum/weakref/bci_to_implant
+	var/obj/item/organ/internal/cyberimp/bci/bci_to_implant
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:284]: If you are going to change this from a weakref you must also implement `/Exited` on bci_implanter so that when the bci_to_implant is removed, this variable becomes null, as to not cause hard deletes.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 4]
```diff
@@ -426,21 +418,20 @@
 	playsound(loc, 'sound/machines/ping.ogg', 30, FALSE)
 
 	var/obj/item/organ/internal/cyberimp/bci/bci_organ = carbon_occupant.getorgan(/obj/item/organ/internal/cyberimp/bci)
-	var/obj/item/organ/internal/cyberimp/bci/bci_to_implant_resolved = bci_to_implant?.resolve()
 
 	if (bci_organ)
 		bci_organ.Remove(carbon_occupant)
 
-		if (isnull(bci_to_implant_resolved))
+		if (!bci_to_implant)
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:425]: Changing these from `isnull` to `!x` is unnecessary and I would rather you kept what it was.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 5]
```diff
@@ -495,6 +486,18 @@
...
+	if (!bci_to_implant)
+		return
+	bci_to_implant.forceMove(drop_location())
+	bci_to_implant = null
+
+/obj/machinery/bci_implanter/dump_inventory_contents(list/subset)
+	// Prevents opening the machine dropping the BCI.
+	// "dump_contents()" still drops the BCI.
+	subset = contents - bci_to_implant
+	return ..()
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:499]:
```suggestion
	return ..(contents - bci_to_implant)
```

More obvious.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 6]
```diff
@@ -459,9 +450,9 @@
 
 	var/mob/living/carbon/carbon_occupant = occupant
 	if (istype(occupant))
-		var/obj/item/organ/internal/cyberimp/bci/existing_bci_organ = carbon_occupant.getorgan(/obj/item/organ/internal/cyberimp/bci)
-		if (isnull(existing_bci_organ) && isnull(bci_to_implant?.resolve()))
-			say("No brain-computer interface inserted, and occupant does not have one. Insert a BCI to implant one.")
+		var/obj/item/organ/internal/cyberimp/bci/bci_organ = carbon_occupant.getorgan(/obj/item/organ/internal/cyberimp/bci)
+		if (!bci_organ && !bci_to_implant)
+			say("No brain-computer interface inserted and the occupant doesn't have one. Insert a BCI to implant one.")
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:455]: Why?

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Some unnecessary changes here

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:296]: Got it.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 3 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:284]: Got it.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:425]: Huh, why?

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 5 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:499]: That... works? (This is copied from skillchip station code.)

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 2 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:301]: Will do.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:425]: Not opposed, though. Will do.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 6 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:455]: Guh, preference. Will revert the text.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:425]: Cuz I wrote it and I like the explicitness

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 5 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:499]: Yeah. `..()` just passes in args, and I always think that's kinda confusing.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:425]: Well if that's the reason I disagree now. A lot of other files use !x and the only thing better than perfect is standardized.

Plus !x is more common in all languages that support it afaik and is more compact.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 5 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:499]: Guh, DM moment. Definitely will do. Should I take care of skillchip station's version too?

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:427]: Okay it's necessary to change it back as it's one of the things we care that we don't care about, like `if (x)` vs `if(x)`

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 6 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:456]: I do want it reverted I think it's needless

Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:456]: Will do.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:427]: I'll do it for this, no problem. However if I make a new feature I'm still going to use the !x method. It's engrained into my coding habits far too deeply.

Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm]: I don't use !x where I could've used isnull

**Comment by RikuTheKiller** [comment]
Finally, all changes have been completed.

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 7]
```diff
@@ -290,21 +290,18 @@
...
-	QDEL_NULL(bci_to_implant)
+	qdel(bci_to_implant)
 	return ..()
 
 /obj/machinery/bci_implanter/examine(mob/user)
-	. = ..()
-
-	if (isnull(bci_to_implant?.resolve()))
+	if (isnull(bci_to_implant))
 		. += span_notice("There is no BCI inserted.")
```
Comment (Watermelon914) [inline code/modules/wiremod/shell/brain_computer_interface.dm:301]: the . operator is not defined here. It's currently null so this will fail. You need to initialize it as a new list at the start of the proc.

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

**Comment by RikuTheKiller** [comment]
Finally???

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 4 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm]: Yet somehow there's !occupant mixed in there a whole bunch.
